lib/logstash/inputs/gelf.rb

```diff
@@ -80,20 +93,91 @@ def run(output_queue)
 
   public
   def stop
-    @udp.close
+    if @protocol.downcase == "tcp"
+      @tcp.close
+    else
+      @udp.close
+    end
   rescue IOError # the plugin is currently shutting down, so its safe to ignore theses errors
   end
 
   private
+  def tcp_listener(output_queue)
+    @logger.info("Starting gelf listener (tcp) ...", :address => "#{@host}:#{@port}")
+
+    if @tcp.nil?
+      @tcp = TCPServer.new(@host, @port)
+    end
+
+    while !stop?
+      Thread.new(@tcp.accept) do |client|
+        @logger.debug? && @logger.debug("Gelf (tcp): Accepting connection from:  #{client.peeraddr[2]}:#{client.peeraddr[1]}")
+
+        begin
+          while !client.nil? && !client.eof?
+
+            begin # Read from socket
+              @data_in = client.gets("\u0000")
+            rescue => ex
+              @logger.warn("Gelf (tcp): failed gets from client socket:", :exception => ex, :backtrace => ex.backtrace)
+            end
+
+             if @data_in.nil?
+              @logger.warn("Gelf (tcp): socket read succeeded, but data is nil.  Skipping.")
+              next
+            end
+
+            # data received.  Remove trailing \0
+            @data_in[-1] == "\u0000" && @data_in = @data_in[0...-1]
+            begin # Parse JSON
+              @jsonObj = JSON.parse(@data_in)
+            rescue => ex
+              @logger.warn("Gelf (tcp): failed to parse a message. Skipping: " + @data_in, :exception => ex, :backtrace => ex.backtrace)
+              next
+            end
+
+            begin  # Create event
+              event = LogStash::Event.new(@jsonObj)
+              event.remove("source_host")
+              event["source_host"] = @use_numeric_client_addr && client.addr(:numeric)[3] || client.addr(:hostname)[3]
```

Comment:
`event[]=` is the old api won't be supported in Logstash 5.0. You'll want to use `event.set(field, value)`